gnosis/safe-react
# refactor: useEstimateTransactionGas hook

## What it solves
Resolves #3499 

## How this PR fixes it

With the `TxModalWrapper` refactor, the hook is only being used in one place. We only estimate gas now if a transaction is being executed.

- Split up `safeTxGasEstimation` and `gasEstimation`
- Move `isCreation` and `isOffchainSignature` out of the hook
- Remove gas estimation for case `isOffChainSignature && !isCreation`
- Only do `gasEstimation` if transaction is executing
- Call `estimateGasForTransactionExecution` instead of `estimateTransactionGasLimit`
- Remove usage of `estimateGasForTransactionApproval`
- Remove unused `manualSafeNonce`, `providerName` from custom hook dependencies
- Remove `manualSafeNonce` from `TxModalWrapper`
- Pass `canTxExecute` instead of declaring it twice with different params
- Rename `canTxExecute` to `isExecution`
- Rename `gasEstimation` to `safeTxGasEstimation`

## Added unit tests

Unit tests for `useEstimateTransactionGas`:

- Should return `manualGasPrice` if it exists instead of estimation
- Should return `manualMaxPrioFee` if it exists instead of estimation
- Should return `manualGasLimit` if it exists instead of estimation
- Should return failure state if `checkTransactionExecution` is false
- Should return failure state and default estimations if `getFeesPerGas` throws
- Should return failure state and default estimations if `estimateGasForTransactionExecution` throws
- Should return failure state and default estimations if `checkTransactionExecution` throws
- Should return failure state and default estimations if `calculateGasPrice` throws
- Should return success state if `checkTransactionExecution` is true
- Should return `gasMaxPrioFee = 0` if pre EIP-1559

Unit tests for `estimateSafeTxGas`:

- Should return 0 if feature flag `SAFE_TX_GAS_OPTIONAL` is set for safe version

## How to test it

All transaction types and safe owner setups need to be tested

- Send Funds in a 1/n safe and execute immediately
- Send Funds in a 1/n safe and queue
- Send Funds in a 1/n safe and execute from queue
- Send Funds in a 2+/n safe and queue
- Send Funds in a 2+/n safe and confirm and execute
- Send Funds in a 2+/n safe and confirm
- Send Funds in a 2+/n safe and execute from queue
- Send Funds in a 2+/n safe and execute as a non-owner
- ...
